### PR TITLE
Auto emulation rewrite step

### PR DESCRIPTION
**Pull request by msoeken**
This PR implements a rewrite step that can link an operation `Op`  to a set of pairs of an alternative operations `AltOp` and a simulator `Sim`, such that `AltOp` is executed as a replacement for `Op` when being invoked in `Sim`.

For example, we could write:

```qsharp
@SubstitutableOnTarget("Microsoft.Quantum.Intrinsic.CCNOT", "ToffoliSimulator")
operation ApplyAnd(control1 : Qubit, control2 : Qubit, target : Qubit) : Unit is Adj+Ctl {...}
```

to replace calls to `ApplyAnd` with `CCNOT` when being called in `ToffoliSimulator`.

This PR can solve the issue raised in #24.

To be done:
* [x] Add integration test project (which loads the rewrite step as Q# reference)
* [x] NuGet packaging
* [x] API review

### CODE REVIEW

**Comment by msoeken** [comment]
@cgranade, I think we need to pass this through API review.

**Comment by msoeken** [comment]
I could use some help with how to package this as a NuGet package.

**Comment by msoeken** [comment]
@bettinaheim @swernli I incorporated the API changes from Microsoft/QuantumLibraries#426 and also added packaging. This PR is ready for review now. Thanks.

**Comment by bettinaheim** [comment]
> @bettinaheim @swernli I incorporated the API changes from [microsoft/QuantumLibraries#426](https://github.com/microsoft/QuantumLibraries/pull/426) and also added packaging. This PR is ready for review now. Thanks.

Lgtm, but I haven't looked at the PR in full detail (e.g. checking for sufficient test coverage etc).

File: src/Simulation/AutoSubstitution/RewriteStep.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                    {
+                        diagnostics.Add(new IRewriteStep.Diagnostic {
+                            Severity = DiagnosticSeverity.Error,
+                            Message = $"AutoSubstitution: name of alternative operation in {key.Namespace}.{key.Name} must be completely specified (including namespace)",
+                            Stage = IRewriteStep.Stage.Transformation
+                        });
+                        return false;
+                    }
+
+                    var qualifiedName = new QsQualifiedName(alternativeOperation.Substring(0, period), alternativeOperation.Substring(period + 1));
```
Comment (cgranade) [inline src/Simulation/AutoSubstitution/RewriteStep.cs:103]: To clarify, does this restrict the substitution target and alternative operation to be in the same namespace? Can `alternativeOperation` contain `.` characters to indicate a subnamespace of that namespace?

File: src/Simulation/AutoSubstitution/RewriteStep.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            // global callables
+            var globalCallables = compilation.Namespaces.GlobalCallableResolutions();
+
+            // collect all callables that have an substitution attribute
+            var globals = globalCallables.Where(p => p.Value.Source.CodeFile.EndsWith(".qs"))
+                                         .Where(p => p.Value.Attributes.Any(HasSubstitutionAttribute));
+
+            if (!globals.Any())
+            {
+                diagnostics.Add(new IRewriteStep.Diagnostic {
```
Comment (cgranade) [inline src/Simulation/AutoSubstitution/RewriteStep.cs:58]: Formatting nit, but braces are typically on the next line for initializer syntax.
```suggestion
                diagnostics.Add(new IRewriteStep.Diagnostic
                {
```

**Review by cgranade — APPROVED** [review]
Had a few last comments and thoughts, but looks good to me!

File: src/Simulation/AutoSubstitution.Tests/CodeGenerationTests.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            var path = CreateNewTemporaryPath();
+            step.AssemblyConstants[AssemblyConstants.OutputPath] = path;
+
+            var compilation = CreateCompilation(Path.Combine("TestFiles", "Core.qs"), "Substitution.qs", Path.Combine("TestFiles", $"{fileName}.qs"));
+
+            Assert.True(step.Transformation(compilation, out var transformed));
+            var generatedFileName = Path.Combine(path, "__AutoSubstitution__.g.cs");
+            Assert.True(File.Exists(generatedFileName));
+
+            // uncomment, when creating files with expected content
```
[the remark this answers is not in the document]
Comment (msoeken) [reply src/Simulation/AutoSubstitution.Tests/CodeGenerationTests.cs:45]: Yes, this is helpful in when creating new unit tests. I make this comment clearer. I saw something similar in qsharp-compiler unit tests.

File: src/Simulation/AutoSubstitution/RewriteStep.cs (hunk 1 above)
Comment (msoeken) [reply src/Simulation/AutoSubstitution/RewriteStep.cs:103]: They can be in different namespaces. Here, I am just splitting the string into namespace + operation name and construct a `QsQualifiedName` instance from the two parts.

File: src/Simulation/AutoSubstitution.Tests/CodeGenerationTests.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            Assert.False(step.Transformation(compilation, out var transformed));
+            Assert.Equal(2, step.GeneratedDiagnostics.Count());
+            Assert.Equal(CodeAnalysis.DiagnosticSeverity.Error, step.GeneratedDiagnostics.Last().Severity);
+        }
+
+        private QsCompilation CreateCompilation(params string[] fileNames)
+        {
+            var mgr = new CompilationUnitManager();
+            var files = CreateFileManager(fileNames);
+            mgr.AddOrUpdateSourceFilesAsync(files).Wait();
```
[the remark this answers is not in the document]
Comment (msoeken) [reply src/Simulation/AutoSubstitution.Tests/CodeGenerationTests.cs:70]: This function is quite deep in the call stack and I decided to not address this comment in favor of avoiding a large change.

**Pull request merged**